File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/common/widget/grid/MyGrid.java
package com.voiceassist.lixinyu.voiceassist.common.widget.grid;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.voiceassist.lixinyu.voiceassist.utils.KGLog;

/**
 * Created by lilidan on 2018/1/22.
 */

public class MyGrid extends LinearLayout {

    private Context mContext;

    private BaseGridAdapter mAdapter;


    public MyGrid(Context context) {
        super(context);
        init(context);
    }

    public MyGrid(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public MyGrid(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {

        this.mContext = context;

        ViewGroup.LayoutParams lp = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        setLayoutParams(lp);
        setOrientation(VERTICAL);
    }

    public void setAdapter(BaseGridAdapter adapter) {
        this.mAdapter = adapter;

        notifyView();
    }

    public void notifyView() {
        int count = mAdapter.getCount();
        int columCount = mAdapter.getColumCount();

        int lineCount = count % columCount == 0 ? count / columCount : count / columCount + 1;
        int lineHeight = getHeight() / lineCount;

        int width = getWidth();
        int itemWidth = width / columCount;

        KGLog.d("aaa", "width--->" + width);
        KGLog.i("aaa", "itemWidth--->" + itemWidth);
        KGLog.v("aaa", "lineHeight--->" + lineHeight);
        KGLog.w("aaa", "getHeight()--->" + getHeight());

        LinearLayout ll = null;
        for (int i = 0; i < count; i++) {
            if (i % columCount == 0) {
                ll = new LinearLayout(mContext);
                ll.setOrientation(HORIZONTAL);
                LayoutParams lpLine = new LayoutParams(LayoutParams.MATCH_PARENT, lineHeight);
                addView(ll, lpLine);
            }


            LayoutParams lpColum = new LayoutParams(itemWidth, LayoutParams.MATCH_PARENT);
            ll.addView(mAdapter.getView(i), lpColum);
        }
    }
}

File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/common/widget/grid/MyGrid.kt
package com.voiceassist.lixinyu.voiceassist.common.widget.grid

import android.content.Context
import android.util.AttributeSet
import android.view.ViewGroup
import android.widget.LinearLayout

import com.voiceassist.lixinyu.voiceassist.utils.KGLog

/**
 * Created by lilidan on 2018/1/22.
 */

class MyGrid : LinearLayout {

    private var mContext: Context? = null

    private var mAdapter: BaseGridAdapter? = null


    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    private fun init(context: Context) {

        this.mContext = context

        val lp = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        layoutParams = lp
        orientation = LinearLayout.VERTICAL
    }

    fun setAdapter(adapter: BaseGridAdapter) {
        this.mAdapter = adapter

        notifyView()
    }

    fun notifyView() {
        val count = mAdapter!!.count
        val columCount = mAdapter!!.columCount

        val lineCount = if (count % columCount == 0) count / columCount else count / columCount + 1
        val lineHeight = height / lineCount

        val width = width
        val itemWidth = width / columCount

        KGLog.d("aaa", "width--->$width")
        KGLog.i("aaa", "itemWidth--->$itemWidth")
        KGLog.v("aaa", "lineHeight--->$lineHeight")
        KGLog.w("aaa", "getHeight()--->$height")

        var ll: LinearLayout? = null
        for (i in 0 until count) {
            if (i % columCount == 0) {
                ll = LinearLayout(mContext)
                ll.orientation = LinearLayout.HORIZONTAL
                val lpLine = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, lineHeight)
                addView(ll, lpLine)
            }


            val lpColum = LinearLayout.LayoutParams(itemWidth, LinearLayout.LayoutParams.MATCH_PARENT)
            ll!!.addView(mAdapter!!.getView(i), lpColum)
        }
    }
}
